### PR TITLE
client: Switch from jQuery.ajax to fetch

### PR DESCRIPTION
**Pull request by jtojnar**
This will bring us one step closer to removing jQuery.

Unfortunately the native promises lack built-in support for aborting so we need to somewhat clumsily return `AbortController` alongside them.

While at it, I also got rid of the jQuery Deferreds so the only remaining use of jQuery is the DOM manipulation. Hopefully, it will not remain for long.

### CODE REVIEW

**Comment by jtojnar** [comment]
@niol will you be available to review this some time? I can also guide you through the Parcel-based build from #1137 if you need to bring up to speed.

File: assets/js/selfoss-db.js [hunk 1]
```diff
@@ -243,57 +248,52 @@ selfoss.dbOnline = {
      */
     reloadList: function() {
         if (selfoss.activeAjaxReq !== null) {
-            selfoss.activeAjaxReq.abort();
+            selfoss.activeAjaxReq.controller.abort();
```
[the remark this answers is not in the document]
Comment (niol) [reply assets/js/selfoss-db.js:251]: This triggers everytime `reloadList` gets called.

**Review by niol — COMMENTED** [review]
I cannot add a source.

File: assets/js/selfoss-db.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (jtojnar) [reply assets/js/selfoss-db.js:251]: Hmm, looks like `liftToPromiseField` does not quite work after the latest refactoring. It drops the `controller` field from the object:

```
> selfoss.ajax.get('sources')
Object { promise: Promise { "pending" } }

> selfoss.ajax.fetch('sources')
Object { controller: AbortController, promise: Promise { "pending" } }
```

[the remark this answers is not in the document]
Comment (jtojnar) [reply assets/js/selfoss-db.js:251]: Looks like it was actually caused by the functions from `fetch-friends` library. Not exactly sure why, but I created our own `selfoss.ajax.option` function and it seems to work.

**Comment by jtojnar** [comment]
Thanks for checking it out. It should be fixed now.

The _add source_ button not functioning was caused by forgetting to call `.text()` on the `fetch`’s `Response`. I checked the rest of the code and found only two other places where we expect HTML instead of JSON and those called it correctly. Hopefully, I did not miss any.

I also had to fix the `Sources\Write` controller to handle missing `tags` field since the new field does not send it when there are no tags.

**Comment by niol** [comment]
With offline storage enabled, the client does not chain load items and stops after `items_perpage`.

**Comment by niol** [comment]
An then it seems to load the same entries over and over again.

**Comment by niol** [comment]
I I shut down the server, it keeps the ajax requests unfinished although the "Connection refused" should terminate them immediately.

**Comment by jtojnar** [comment]
If I manually trigger `selfoss.dbOffline.sendNewStatuses()`, it seems to fetch another page. I am not sure what is supposed to trigger the fetch of the next page.

**Comment by niol** [comment]
when `selfoss.dbOffline.newerEntriesMissing` is true.

**Comment by jtojnar** [comment]
That seems to be true here but no luck.

**Comment by niol** [comment]
I would try the following but I'm stuck with a bundler error.

```diff
--- a/assets/js/selfoss-db.js
+++ b/assets/js/selfoss-db.js
@@ -157,16 +157,19 @@ selfoss.dbOnline = {
                     selfoss.dbOffline.storeEntries(data.newItems)
                         .then(function() {
                             selfoss.dbOffline.storeLastUpdate(dataDate);
-
                             selfoss.dbOnline._syncDone();
-
-                            // fetch more if server has more
-                            if (selfoss.dbOffline.newerEntriesMissing) {
-                                selfoss.dbOnline.sync();
-                            }
                         });
                 }
 
+                if (selfoss.dbOffline.newerEntriesMissing
+                    || selfoss.dbOffline.needsSync) {
+                    // There are still new items to fetch
+                    // or statuses to send
+                    syncing.request.promise.then(function () {
+                        selfoss.dbOffline.sendNewStatuses();
+                    });
+                }
+
                 if ('itemUpdates' in data) {
                     // refresh entry statuses in db and dequeue queued
                     // statuses but do not calculate stats as they are taken
```

**Comment by jtojnar** [comment]
Yeah, the bundlers are somewhat finicky in my experience. If I do `git rebase` with `npm run dev` running it breaks horribly, and I need to delete `assets/.cache` and restart it. OR sometimes on a syntax error, it is just swallowed and I need to restart it to see the error message. Unfortunately, when I tried Webpack it did not seem much better.

That seems to work with a minor tweak:

```patch
--- a/assets/js/selfoss-db.js
+++ b/assets/js/selfoss-db.js
@@ -159,14 +159,18 @@ selfoss.dbOnline = {
                             selfoss.dbOffline.storeLastUpdate(dataDate);
 
                             selfoss.dbOnline._syncDone();
-
-                            // fetch more if server has more
-                            if (selfoss.dbOffline.newerEntriesMissing) {
-                                selfoss.dbOnline.sync();
-                            }
                         });
                 }
 
+                if (selfoss.dbOffline.newerEntriesMissing
+                    || selfoss.dbOffline.needsSync) {
+                    // There are still new items to fetch
+                    // or statuses to send
+                    selfoss.dbOnline.syncing.request.promise.then(function () {
+                        selfoss.dbOffline.sendNewStatuses();
+                    });
+                }
+
                 if ('itemUpdates' in data) {
                     // refresh entry statuses in db and dequeue queued
                     // statuses but do not calculate stats as they are taken
```

Thanks.

**Comment by jtojnar** [comment]
But that patch triggers a sync loop you mention, the last page does not seem to get picked up.

**Comment by niol** [comment]
The sync loop seems to be specific to sqlite and how we compare dates as simple strings.

We should either use specific syntax for sqlite queries or ensure compared dates share the same timezone.

For instance, the following seems to be fixing the problem.
```diff
--- a/src/controllers/Items/Sync.php
+++ b/src/controllers/Items/Sync.php
@@ -56,11 +56,12 @@ class Sync {
         }

         $since = new \DateTime($params['since']);
+        $since->setTimeZone(new \DateTimeZone(date_default_timezone_get()));

         $last_update = new \DateTime($this->itemsDao->lastUpdate());

         $sync = [
-            'lastUpdate' => $last_update->format('Y-m-d H:i:s'),
+            'lastUpdate' => $last_update->format(\DateTime::ATOM),
         ];

         $sinceId = 0;
@@ -73,6 +74,7 @@ class Sync {
                     // only send 1 day worth of items
                     $notBefore = new \DateTime();
                     $notBefore->sub(new \DateInterval('P1D'));
+                    $notBefore->->setTimeZone(new \DateTimeZone(date_default_timezone_get()));
                 }

                 $itemsHowMany = $f3->get('items_perpage');
```

**Review by niol — CHANGES_REQUESTED** [review]

**Pull request merged**